# BookCycle SQL Analysis

## Project Overview

This project analyzes BookCycle's books, customers, and transaction data using SQL.

The analysis demonstrates practical SQL skills including:

- Data filtering
- Aggregation
- Sorting
- JOINs
- Subqueries
- Common Table Expressions (CTEs)
- Window functions

## Business Objective

The goal is to use BookCycle's data to identify patterns in books, transactions, customers, and sales that can support data-driven business decisions.  

## Dataset

The BookCycle database contains three main tables:

### 1. Books
Contains information about BookCycle's book inventory, including:
- Book ID
- Title
- Author
- ISBN
- Genre
- Condition
- Purchase price
- List price
- Date acquired
- Current location
- Quantity

### 2. Customers
Contains customer information, including:
- Customer ID
- Join date
- Membership status
- ZIP code
- Birth year
- Preferred store

### 3. Transactions
Contains sales transaction information, including:
- Transaction ID
- Date and time
- Store location
- Customer ID
- Book ID
- Sale price
- Payment method
- Online/offline status

## Data Preparation

The BookCycle data was loaded into a SQLite database before performing the SQL analysis. The database contains three related tables: books, customers, and transactions.

## SQL Skills Demonstrated

This project demonstrates the following SQL skills:

| SQL Skill | Examples Used |
|---|---|
| Filtering | `WHERE` |
| Aggregation | `SUM()`, `AVG()`, `COUNT()` |
| Grouping | `GROUP BY`, `HAVING` |
| Sorting | `ORDER BY` |
| Joins | `JOIN` across multiple tables |
| Subqueries | Average-price and high-value transaction analysis |
| CTEs | Multi-step sales analysis |
| Window Functions | `ROW_NUMBER()`, `PARTITION BY` |

## SQL Analysis

### 1. Find Books in the Classic Fiction Genre

**Business question:**  
Which books belong to the Classic Fiction genre?

In [5]:
query = """
SELECT *
FROM books
WHERE genre = 'Classic Fiction';
"""

execute_query(query)

,book_id,title,author,isbn,genre,condition,purchase_price,list_price,date_acquired,current_location,quantity
0,B1001,A Christmas Carol,Charles Dickens,9780141324524,Classic Fiction,Good,5.5,8.99,2023-01-15,Suburban,2
1,B1002,A Farewell to Arms,Ernest Hemingway,9780684801469,Classic Fiction,Very Good,7.0,11.99,2023-01-15,University,3
2,B1003,A Tale of Two Cities,Charles Dickens,9780141439600,Classic Fiction,Fair,4.5,7.99,2023-01-16,Downtown,2
3,B1004,Adventures of Huckleberry Finn,Mark Twain,9780142437179,Classic Fiction,Good,6.0,9.99,2023-01-16,University,4
4,B1005,Agnes Grey,Anne Bronte,9780140432107,Classic Fiction,Fair,4.0,7.99,2023-01-17,Suburban,1
...,...,...,...,...,...,...,...,...,...,...,...
68,B1093,The Trial,Franz Kafka,9780141182902,Classic Fiction,Good,6.5,10.99,2023-03-02,University,3
69,B1095,The Wind in the Willows,Kenneth Grahame,9780143039099,Classic Fiction,Good,5.5,9.99,2023-03-03,Suburban,2
70,B1097,Treasure Island,Robert Louis Stevenson,9780141321011,Classic Fiction,Fair,4.5,7.99,2023-03-04,Suburban,2
71,B1098,Ulysses,James Joyce,9780141182803,Classic Fiction,Very Good,8.0,13.99,2023-03-04,University,3


### 2. Find Books Above the Average List Price

**Business question:**  
Which books have a list price higher than the average list price of all books?

In [6]:
query = """
SELECT *
FROM books
WHERE list_price > (
    SELECT AVG(list_price)
    FROM books
);
"""

execute_query(query)

,book_id,title,author,isbn,genre,condition,purchase_price,list_price,date_acquired,current_location,quantity
0,B1002,A Farewell to Arms,Ernest Hemingway,9780684801469,Classic Fiction,Very Good,7.0,11.99,2023-01-15,University,3
1,B1007,Anna Karenina,Leo Tolstoy,9780143035008,Classic Fiction,Good,7.0,11.99,2023-01-18,University,3
2,B1008,Anne of Green Gables,L.M. Montgomery,9780141321598,Classic Fiction,Like New,8.0,12.99,2023-01-18,Suburban,2
3,B1013,Brave New World,Aldous Huxley,9780060850524,Classic Fiction,Very Good,7.5,12.99,2023-01-21,University,4
4,B1016,Crime and Punishment,Fyodor Dostoevsky,9780140449136,Classic Fiction,Like New,8.5,13.99,2023-01-22,University,4
5,B1019,Don Quixote,Miguel de Cervantes,9780142437230,Classic Fiction,Good,7.0,11.99,2023-01-24,Downtown,2
6,B1022,Fahrenheit 451,Ray Bradbury,9780143036142,Classic Fiction,Like New,8.0,12.99,2023-01-25,University,4
7,B1024,Frankenstein,Mary Shelley,9780141439471,Classic Fiction,Very Good,7.0,11.99,2023-01-26,University,3
8,B1029,Iliad,Homer,9780140275360,Classic Poetry,Like New,8.0,13.99,2023-01-29,University,4
9,B1031,Invisible Man,Ralph Ellison,9780679732761,Classic Fiction,Very Good,7.0,11.99,2023-01-30,University,4


### 3. Total List Price by Location

**Business question:**  
What is the total list price of books at each current location?

In [7]:
query = """
SELECT
    current_location,
    SUM(list_price) AS total_value
FROM books
GROUP BY current_location
ORDER BY total_value DESC;
"""

execute_query(query)

,current_location,total_value
0,University,634.45
1,Downtown,293.72
2,Suburban,172.83


### 4. Find High-Value Transactions

**Business question:**  
Which transactions have a sale price higher than the average transaction sale price?

In [5]:
query = """
SELECT *
FROM transactions
WHERE sale_price > (
    SELECT AVG(sale_price)
    FROM transactions
)
ORDER BY sale_price DESC;
"""

execute_query(query)

,transaction_id,date_time,store_location,customer_id,book_id,sale_price,payment_method,is_online
0,T1004,2023-01-15 11:30:15,University,C1012,B1071,13.99,credit,1
1,T1032,2023-01-19 11:40:45,University,C1023,B1096,13.99,credit,0
2,T1039,2023-01-20 13:30:12,University,C1023,B1071,13.99,credit,1
3,T1051,2023-01-23 09:15:22,University,C1089,B1071,13.99,credit,0
4,T1062,2023-01-24 14:40:45,University,C1045,B1098,13.99,credit,0
5,T1070,2023-01-25 15:55:48,University,C1045,B1071,13.99,debit,0
6,T1093,2023-01-30 10:45:33,University,C1023,B1071,13.99,credit,1
7,T1005,2023-01-15 13:45:22,University,C1034,B1075,12.99,debit,0
8,T1007,2023-01-15 15:30:45,University,C1089,B1047,12.99,credit,0
9,T1016,2023-01-17 09:30:22,University,C1012,B1082,12.99,debit,1


### 5. Analyze Customer Purchases

**Business question:**  
Which customers have made more than 5 purchases, and how much have they spent in total?

In [6]:
query = """
SELECT
    c.customer_id,
    COUNT(t.transaction_id) AS total_purchases,
    SUM(t.sale_price) AS total_amount_spent
FROM customers AS c
JOIN transactions AS t
    ON c.customer_id = t.customer_id
GROUP BY c.customer_id
HAVING COUNT(t.transaction_id) > 5
ORDER BY total_amount_spent DESC;
"""

execute_query(query)

,customer_id,total_purchases,total_amount_spent
0,C1023,13,155.87
1,C1089,13,154.87
2,C1012,14,153.86
3,C1034,13,146.87
4,C1045,12,136.88
5,C1078,12,132.88
6,C1067,12,119.88
7,C1056,11,113.89


### 6. Combine Customer, Transaction, and Book Data

**Business question:**  
What books did customers purchase, when did they purchase them, where did the purchase occur, and how much did they pay?

In [5]:
query = """
SELECT
    c.customer_id,
    c.preferred_store,
    t.transaction_id,
    t.date_time,
    t.store_location,
    b.book_id,
    b.title,
    b.genre,
    t.sale_price,
    t.payment_method,
    t.is_online
FROM customers AS c
JOIN transactions AS t
    ON c.customer_id = t.customer_id
JOIN books AS b
    ON t.book_id = b.book_id
ORDER BY t.date_time;
"""

execute_query(query)

,customer_id,preferred_store,transaction_id,date_time,store_location,book_id,title,genre,sale_price,payment_method,is_online
0,C1045,Suburban,T1001,2023-01-15 09:23:45,University,B1055,Pride and Prejudice,Classic Fiction,11.99,credit,0
1,C1023,University,T1002,2023-01-15 10:15:22,Downtown,B1032,Jane Eyre,Classic Fiction,10.99,debit,0
2,C1078,University,T1003,2023-01-15 10:45:33,Suburban,B1036,Little Women,Classic Fiction,11.99,cash,0
3,C1012,University,T1004,2023-01-15 11:30:15,University,B1071,The Catcher in the Rye,Classic Fiction,13.99,credit,1
4,C1034,University,T1005,2023-01-15 13:45:22,University,B1075,The Great Gatsby,Classic Fiction,12.99,debit,0
...,...,...,...,...,...,...,...,...,...,...,...
95,C1078,University,T1096,2023-01-30 14:45:22,University,B1094,The Waste Land,Classic Poetry,9.99,debit,0
96,C1067,Suburban,T1097,2023-01-31 09:25:45,University,B1082,The Odyssey,Classic Poetry,12.99,credit,0
97,C1056,Downtown,T1098,2023-01-31 10:40:33,Downtown,B1003,A Tale of Two Cities,Classic Fiction,7.99,cash,0
98,C1089,Downtown,T1099,2023-01-31 11:50:12,University,B1075,The Great Gatsby,Classic Fiction,12.99,credit,1


### 7. Top-Selling Book in Each Genre at Each Store

**Business question:**  
What is the top-selling book in each genre at each BookCycle store?

In [6]:
query = """
WITH book_sales AS (
    SELECT
        t.store_location,
        b.genre,
        b.book_id,
        b.title,
        COUNT(t.transaction_id) AS total_sold
    FROM transactions AS t
    JOIN books AS b
        ON t.book_id = b.book_id
    GROUP BY
        t.store_location,
        b.genre,
        b.book_id,
        b.title
),
ranked_books AS (
    SELECT
        store_location,
        genre,
        book_id,
        title,
        total_sold,
        ROW_NUMBER() OVER (
            PARTITION BY store_location, genre
            ORDER BY total_sold DESC
        ) AS rank
    FROM book_sales
)
SELECT
    store_location,
    genre,
    book_id,
    title,
    total_sold
FROM ranked_books
WHERE rank = 1
ORDER BY
    store_location,
    genre;
"""

execute_query(query)

,store_location,genre,book_id,title,total_sold
0,Downtown,Classic Fiction,B1001,A Christmas Carol,4
1,Suburban,Classic Fiction,B1095,The Wind in the Willows,5
2,Suburban,Classic Mystery,B1064,The Adventures of Sherlock Holmes,4
3,University,Classic Drama,B1027,Hamlet,4
4,University,Classic Fiction,B1071,The Catcher in the Rye,5
5,University,Classic Non-Fiction,B1099,Walden,2
6,University,Classic Philosophy,B1084,The Prince,2
7,University,Classic Poetry,B1082,The Odyssey,4


### 8. Customers with More Than 5 Purchases

**Business question:**  
Which customers have made more than 5 purchases, and how much have they spent in total?

In [7]:
query = """
SELECT
    c.customer_id,
    COUNT(t.transaction_id) AS total_purchases,
    SUM(t.sale_price) AS total_amount_spent
FROM customers AS c
JOIN transactions AS t
    ON c.customer_id = t.customer_id
GROUP BY c.customer_id
HAVING COUNT(t.transaction_id) > 5
ORDER BY total_amount_spent DESC;
"""

execute_query(query)

,customer_id,total_purchases,total_amount_spent
0,C1023,13,155.87
1,C1089,13,154.87
2,C1012,14,153.86
3,C1034,13,146.87
4,C1045,12,136.88
5,C1078,12,132.88
6,C1067,12,119.88
7,C1056,11,113.89


## 4. Key Insights

### Book Inventory
- The University location has the highest total list price of books at **634.45**, followed by Downtown at **293.72** and Suburban at **172.83**.
- The analysis identified books in the **Classic Fiction** genre and books priced above the overall average list price.

### Transaction Analysis
- The high-value transaction analysis identifies transactions with sale prices above the overall transaction average.
- The highest sale prices in the returned high-value transactions reached **13.99**.

### Customer Analysis
- Several customers made more than 5 purchases.
- Customer **C1023** had the highest total spending among customers with more than 5 purchases, with **13 purchases totaling 155.87**.
- Customer **C1012** made the highest number of purchases in this group, with **14 purchases**.

### Store and Genre Analysis
- The top-selling book varies by store and genre.
- For example, at the University store, **The Catcher in the Rye** was the top-selling Classic Fiction book with **5 sales**.

## 5. Business Recommendations

Based on the SQL analysis, BookCycle could:

1. **Monitor inventory value by location**  
   Review inventory allocation across stores and identify opportunities to balance book availability and inventory value.

2. **Identify high-value and frequently purchasing customers**  
   Consider targeted customer engagement or loyalty initiatives for customers with higher purchase frequency.

3. **Use store-level and genre-level sales patterns**  
   Use top-selling books by store and genre to support inventory planning and stocking decisions.

4. **Monitor high-value transactions**  
   Track higher-value purchases to better understand customer purchasing behavior and identify opportunities for additional sales.

5. **Continue analyzing sales trends**  
   Combine transaction data with time-based analysis to identify seasonal and monthly purchasing patterns.

## 6. Conclusion

This project demonstrates how SQL can be used to transform relational BookCycle data into business insights.

The analysis applied filtering, aggregation, subqueries, JOINs, CTEs, and window functions to investigate books, transactions, customers, and store-level sales patterns.

These techniques can support data-driven decisions related to inventory management, customer engagement, and sales planning.

## Data Setup

The analysis uses three related BookCycle tables:

- **books** — book inventory and pricing information
- **customers** — customer and membership information
- **transactions** — purchase and sales transaction information

### Connect to the SQLite Database



In [2]:
import sqlite3
import pandas as pd

# Connect to the BookCycle database
conn = sqlite3.connect('bookcycle.db')
cursor = conn.cursor()

# Function to execute SQL queries and display results
def execute_query(query):
    df = pd.read_sql_query(query, conn)
    display(df)

### Create the Books Table

In [3]:
# Create the books table
create_table_query = """
CREATE TABLE IF NOT EXISTS books (
    book_id TEXT PRIMARY KEY,  
    title TEXT,
    author TEXT,
    isbn TEXT,
    genre TEXT,
    condition TEXT,
    purchase_price REAL,
    list_price REAL,
    date_acquired TEXT, 
    current_location TEXT,
    quantity INTEGER
)
"""

cursor.execute(create_table_query)
conn.commit()
print("Table 'books' created successfully (or already exists).")

Table 'books' created successfully (or already exists).


### Import Book Data

In [4]:
# Read the CSV into a pandas DataFrame
df = pd.read_csv("books.csv")

# Write the DataFrame to the database table
df.to_sql("books", conn, if_exists='replace', index=False)  # if_exists='append' if adding

conn.commit()
print("CSV data imported to 'books' table successfully.")


CSV data imported to 'books' table successfully.
